### PR TITLE
Conjugate heat transfer for segregated solvers in MOOSE-NS 

### LINKED ISSUE #31203: Need conjugate heat transfer capabilities for the segregated solver in MOOSE-NS.
### Motivation

For engineering applications CHT is an essential tool. Would be nice to have it in MOOSE-NS.

### Design

Add boundary conditions, and modifying the loop.

### Impact

Will enable the direct computation of heat-transfer coefficients.

### PR DESCRIPTION
**Pull request by anshchaube**
Closes #31203
cc @grmnptr

### CODE REVIEW

File: framework/src/linearfvkernels/LinearFVFluxKernel.C [hunk 1]
```diff
@@ -193,6 +193,16 @@ LinearFVFluxKernel::singleSidedFaceArg(const FaceInfo * fi,
...
 }
 
+Real
+LinearFVFluxKernel::computeBoundaryFlux(const LinearFVBoundaryCondition & bc)
+{
+  const auto elem_info = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM)
+                             ? _current_face_info->elemInfo()
+                             : _current_face_info->neighborInfo();
+  return computeBoundaryMatrixContribution(bc) * _var.getElemValue(*elem_info, determineState()) -
+         computeBoundaryRHSContribution(bc);
```
Comment (GiudGiud) [inline framework/src/linearfvkernels/LinearFVFluxKernel.C:203]: that is nice and always true

File: modules/navier_stokes/doc/content/design/linear_fv_cht.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# Conjugate Heat Transfer (CHT) Capability
+
+This summarizes the design and details of the conjugate heat transfer capabilities
+with the linear finite volume system through the [SIMPLE.md] execitoner.
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/design/linear_fv_cht.md]:
```suggestion
with the linear finite volume system through the [SIMPLE.md] executioner.
```

File: modules/navier_stokes/doc/content/design/linear_fv_cht.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+This capability is activate by specifying a boundary on the [!param](/Executioner/SIMPLE/bcht_interfaces)
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/design/linear_fv_cht.md]:
```suggestion
This capability is activated by specifying a boundary on the [!param](/Executioner/SIMPLE/bcht_interfaces)
```

File: modules/navier_stokes/doc/content/design/linear_fv_cht.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+used boundary conditions are compatible or not. In general, we introduced CHT versions of
+common boundary conditions such as [LinearFVRobinCHTBC.md] and [LinearFVDirichletCHTBC.md]
+that are dedicated for CHT applications.
+
+For coupling purposes several new functors are created under the hood:
+
+- +heat_flux_to_solid_*+ (where * is the interface boudary name),
+- +heat_flux_to_fluid_*+ (where * is the interface boudary name),
+- +interface_temperature_solid_*+ (where * is the interface boudary name),
+- +interface_temperature_fluid_*+ (where * is the interface boudary name),
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/design/linear_fv_cht.md]:
```suggestion
- +heat_flux_to_solid_*+ (where * is the interface boundary name),
- +heat_flux_to_fluid_*+ (where * is the interface boundary name),
- +interface_temperature_solid_*+ (where * is the interface boundary name),
- +interface_temperature_fluid_*+ (where * is the interface boundary name),
```

File: modules/navier_stokes/doc/content/design/linear_fv_cht.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+2. **Continuity of Conductive Flux at the Interface**
+   \begin{equation}
+       q_\mathrm{f,wall} = -q_\mathrm{s,wall}
+   \end{equation}
+
+## Coupling Methods
+
+The currenty recommended methods utilize [LinearFVDirichletCHTBC.md] and
+[LinearFVRobinCHTBC.md] in the two different ways listed below. The Robin BC can also
+emulate a neuamnn BC.
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/design/linear_fv_cht.md]:
```suggestion
emulate a neumann BC by setting ....
```

Comment (GiudGiud) [reply modules/navier_stokes/doc/content/design/linear_fv_cht.md]: maybe mention same result?

File: modules/navier_stokes/doc/content/design/linear_fv_cht.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+   \end{equation}
+
+2. **Continuity of Conductive Flux at the Interface**
+   \begin{equation}
+       q_\mathrm{f,wall} = -q_\mathrm{s,wall}
+   \end{equation}
+
+## Coupling Methods
+
+The currenty recommended methods utilize [LinearFVDirichletCHTBC.md] and
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/design/linear_fv_cht.md]:
```suggestion
The methods currently recommended for CHT utilize [LinearFVDirichletCHTBC.md] and
```

File: modules/navier_stokes/include/base/NS.h [hunk 7]
```diff
@@ -187,6 +187,16 @@ enum class WallTreatmentEnum
   NEQ = 3
 };
 
+/**
+ * CHT side options, we want to make sure these can be used as integers
+ * so we are avoiding the enum class here.
```
Comment (GiudGiud) [inline modules/navier_stokes/include/base/NS.h:192]: ? the MooseEnum?

File: modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVDirichletCHTBC.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# LinearFVDirichletCHTBC
+
+## Description
+
+This object is a boundary condition used in the Navier-Stokes module that can be used for conjugate heat transfer (CHT) problems.
+Its role is to enforce a fixed Dirichlet value at an interface where heat is exchanged between solid and fluid regions.
+In practice, it is a wrapper around [LinearFVAdvectionDiffusionFunctorDirichletBC.md] with additional content
+allowing error checking in CHT applications. For more information on the design of CHT, click [here](linear_fv_cht.md).
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVDirichletCHTBC.md]:
```suggestion
allowing error checking in CHT applications. For more information on the design of CHT, see [the CHT capability page](linear_fv_cht.md).
```

File: modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVRobinCHTBC.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
+# LinearFVRobinCHTBC
+
+## Description
+
+This object is a boundary condition used in the Navier-Stokes module that can be used for conjugate heat transfer (CHT) problems.
+Its role is to enforce a Robin-type boundary condition at an interface where heat is exchanged between solid and fluid regions.
+It is similar in behavior to [LinearFVAdvectionDiffusionFunctorRobinBC.md] with the
+coefficients in the expression aligned with the thermal-hydraulics fields.
+For more information on the design of CHT, click [here](linear_fv_cht.md).
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVRobinCHTBC.md]:
```suggestion
For more information on the design of CHT, see [the CHT capability page](linear_fv_cht.md).
```

File: modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVRobinCHTBC.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+- heat_flux_to_solid_* (where * is the interface boudary name)
+- heat_flux_to_fluid_* (where * is the interface boudary name)
+
+Furthermore, this boundary condition also requires a surface temperature which
+is typically the surface temperature on the other side of the interface.
+The interface temperatures are also tracked in fuctors and are created within the
+[SIMPLE.md] executioner. The following surface termperatures are created automatically:
+
+- interface_temperature_solid_* (where * is the interface boudary name)
+- interface_temperature_fluid_* (where * is the interface boudary name)
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVRobinCHTBC.md]:
```suggestion
- heat_flux_to_solid_* (where * is the interface boundary name)
- heat_flux_to_fluid_* (where * is the interface boundary name)

Furthermore, this boundary condition also requires a surface temperature which
is typically the surface temperature on the other side of the interface.
The interface temperatures are also tracked in fuctors and are created within the
[SIMPLE.md] executioner. The following surface termperatures are created automatically:

- interface_temperature_solid_* (where * is the interface boundary name)
- interface_temperature_fluid_* (where * is the interface boundary name)
```

File: modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVRobinCHTBC.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/linearfvbcs/LinearFVRobinCHTBC.md:28]: you could say "for the fluid equation, set flux to ... and temperautre to ..." and same for solid equation

File: modules/navier_stokes/include/executioners/LinearAssemblySegregatedSolve.h [hunk 12]
```diff
@@ -12,6 +12,7 @@
 // Moose includes
 #include "RhieChowMassFlux.h"
 #include "SIMPLESolveBase.h"
+#include "CHTHandler.h"
```
Comment (GiudGiud) [inline modules/navier_stokes/include/executioners/LinearAssemblySegregatedSolve.h:15]:
```suggestion

class NS::FV::CHTHandler;
```

File: modules/navier_stokes/src/executioners/LinearAssemblySegregatedSolve.C [hunk 13]
```diff
@@ -11,6 +11,7 @@
 #include "FEProblem.h"
 #include "SegregatedSolverUtils.h"
 #include "LinearSystem.h"
+#include "Executioner.h"
```
Comment (GiudGiud) [inline modules/navier_stokes/src/executioners/LinearAssemblySegregatedSolve.C:15]:
```suggestion
#include "CHTHandler.h"

```

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/gold/cht_neu-dir_out_y_vs_tf_0001.csv [hunk 14]
```diff
@@ -0,0 +1,13 @@
+T_fluid,id,x,y,z
+821.92762844174,0,0.05,1e-09,0
+890.44702151345,0.00090909090909091,0.05,0.00090909190909091,0
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/gold/cht_neu-dir_out_y_vs_tf_0001.csv:3]: why is id fractional??

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    requirement = 'The system shall be able to solve conjugate heat transfer problems using a Neumann-Dirichlet boundary conditions pair '
+    [bottom]
+      type = CSVDiff
+      input = cht_neu-dir_bot.i
+      csvdiff = 'cht_neu-dir_bot_out_y_vs_tf_0001.csv cht_neu-dir_bot_out_y_vs_ts_0001.csv'
+      detail = 'with the heat transfer interface normals matching the principal normal directions.'
+    []
+    [top]
+      type = CSVDiff
+      input = cht_neu-dir_top.i
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests]: do you really need a full input to switch the side? cant you just rotate the geometry?

Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests]: https://mooseframework.inl.gov/source/meshgenerators/SymmetryTransformGenerator.html with plane symmetry would do it?

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/cht_neu-dir.i [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+  momentum_petsc_options_iname = '-pc_type -pc_hypre_type'
+  momentum_petsc_options_value = 'hypre boomeramg'
+  pressure_petsc_options_iname = '-pc_type -pc_hypre_type'
+  pressure_petsc_options_value = 'hypre boomeramg'
+  energy_petsc_options_iname = '-pc_type -pc_hypre_type'
+  energy_petsc_options_value = 'hypre boomeramg'
+  solid_energy_petsc_options_iname = '-pc_type -pc_hypre_type'
+  solid_energy_petsc_options_value = 'hypre boomeramg'
+  print_fields = false
+  max_cht_fpi = 3
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/cht_neu-dir.i:431]: it d be nice to group CHT parameters in the input for readability

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/cht_neu-dir.i [hunk 17]
```diff
@@ -0,0 +1,430 @@
...
+    functor = 0
+    boundary = 'solid_left solid_right'
+  []
+  [outlet_T]
+    type = LinearFVAdvectionDiffusionOutflowBC
+    variable = T_fluid
+    use_two_term_expansion = false
+    boundary = 'fluid_right'
+  []
+
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/cht_neu-dir.i:321]:
```suggestion
```

or group T_solid and T_fluid

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/cht_rob-rob_bot.i [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+    advected_quantity = temperature
+    cp = ${cp}
+    advected_interp_method = ${advected_interp_method}
+    rhie_chow_user_object = 'rc'
+  []
+  [conduction]
+    type = LinearFVDiffusion
+    variable = T_fluid
+    diffusion_coeff = ${k}
+    use_nonorthogonal_correction = true
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/cht_rob-rob_bot.i]:
```suggestion
    use_nonorthogonal_correction = false
```
same everywhere now that missing gradient request is fixed

File: modules/navier_stokes/include/utils/CHTHandler.h [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+  /// The relaxation factors for temperature fields for the CHT boundaries
+  /// first index is solid/fluid second is the interface
+  std::vector<std::vector<Real>> _cht_flux_relaxation_factor;
+
+  /// The relaxation factors for temperature fields for the CHT boundaries
+  /// first index is solid/fluid second is the interface
+  std::vector<std::vector<Real>> _cht_temperature_relaxation_factor;
+
+  /// The relaxation factors for flux fields for the CHT boundaries
+  /// first index is the boundary, the second solid/fluid.
```
Comment (GiudGiud) [inline modules/navier_stokes/include/utils/CHTHandler.h]: update docstring

File: modules/navier_stokes/include/utils/CHTHandler.h [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+  const unsigned int _max_cht_fpi;
+
+  /// Tolerance for heat flux at the CHT interfaces
+  const Real _cht_heat_flux_tolerance;
+
+  /// The relaxation factors for temperature fields for the CHT boundaries
+  /// first index is solid/fluid second is the interface
+  std::vector<std::vector<Real>> _cht_flux_relaxation_factor;
+
+  /// The relaxation factors for temperature fields for the CHT boundaries
```
Comment (GiudGiud) [inline modules/navier_stokes/include/utils/CHTHandler.h:103]:
```suggestion
  /// The relaxation factors for the flux fields for the CHT boundaries
```

File: modules/navier_stokes/src/linearfvbcs/LinearFVRobinCHTBC.C [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+registerMooseObject("NavierStokesApp", LinearFVRobinCHTBC);
+
+InputParameters
+LinearFVRobinCHTBC::validParams()
+{
+  InputParameters params = LinearFVAdvectionDiffusionBC::validParams();
+  params.addRequiredParam<MooseFunctorName>("h", "The convective heat transfer coefficient.");
+  params.addRequiredParam<MooseFunctorName>("incoming_flux",
+                                            "The incoming diffusive flux on the intervace.");
```
Comment (GiudGiud) [inline modules/navier_stokes/src/linearfvbcs/LinearFVRobinCHTBC.C]:
```suggestion
                                            "The incoming diffusive flux on the interface.");
```

File: modules/navier_stokes/src/linearfvbcs/LinearFVRobinCHTBC.C [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+{
+  const auto elem_info = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM)
+                             ? _current_face_info->elemInfo()
+                             : _current_face_info->neighborInfo();
+  const auto state = determineState();
+  auto face = singleSidedFaceArg(_current_face_info);
+  face.face_side = elem_info->elem();
+
+  return _htc(face, state) *
+             (_var.getElemValue(*elem_info, determineState()) - _surface_temperature(face, state)) +
```
Comment (GiudGiud) [inline modules/navier_stokes/src/linearfvbcs/LinearFVRobinCHTBC.C]: umm isnt the normal graident * conductivty equal to all this

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 23]
```diff
@@ -0,0 +1,417 @@
...
+      if (bcs.size() != 1)
+        mooseError("We found multiple or no boundary conditions for solid energy on boundary ",
+                   boundary_name,
+                   " (ID: ",
+                   boundary_id,
+                   "). Make sure you define exactly one for conjugate heat transfer applications!");
+      _cht_boundary_conditions[bd_index][region_index] = bcs[0];
+
+      if (!dynamic_cast<LinearFVCHTBCInterface *>(_cht_boundary_conditions[bd_index][region_index]))
+        mooseError("The celected boundary condition cannot be used with CHT problems! Make sure it "
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]:
```suggestion
        mooseError("The selected boundary condition cannot be used with CHT problems! Make sure it "
```

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 24]
```diff
@@ -0,0 +1,417 @@
...
+CHTHandler::initializeCHTCouplingFields()
+{
+  for (const auto bd_index : index_range(_cht_boundary_ids))
+  {
+    const auto & bd_fi_container = _cht_face_info[bd_index];
+    auto & temperature_container = _boundary_temperature[bd_index];
+
+    for (const auto region_index : make_range(2))
+    {
+      // Can't be const considering we will updat emembers from here
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]:
```suggestion
      // Can't be const considering we will update members from here
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 25]
```diff
@@ -0,0 +1,417 @@
...
+      " conjugate heat transfer simulations. The default value of 1 essentially keeps"
+      " the FPI feature turned off.");
+
+  params.addRangeCheckedParam<Real>(
+      "cht_heat_flux_tolerance",
+      1e-5,
+      "cht_heat_flux_tolerance > 0",
+      "The relative tolerance for terminating conjugate heat transfer iteration.");
+
+  params.addRangeCheckedParam<std::vector<Real>>(
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]: missing range check?

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 26]
```diff
@@ -0,0 +1,417 @@
...
+      1,
+      "max_cht_fpi >= 1",
+      "Number of maximum fixed point iterations (FPI). Currently only applied to"
+      " conjugate heat transfer simulations. The default value of 1 essentially keeps"
+      " the FPI feature turned off.");
+
+  params.addRangeCheckedParam<Real>(
+      "cht_heat_flux_tolerance",
+      1e-5,
+      "cht_heat_flux_tolerance > 0",
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]: upper bound?

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 27]
```diff
@@ -0,0 +1,417 @@
...
+      {},
+      "The relaxation factors for the boundary temperature when being updated on the fluid side.");
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "cht_solid_temperature_relaxation",
+      {},
+      "The relaxation factors for the boundary temperature when being updated on the solid side.");
+
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "cht_fluid_flux_relaxation",
+      {},
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C:60]: same

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 28]
```diff
@@ -0,0 +1,417 @@
...
+    const auto & bd_fi_container = _cht_face_info[bd_index];
+
+    // We enter the face loop to update the coupling fields
+    for (const auto & fi : bd_fi_container)
+    {
+      other_kernel->setupFaceData(fi);
+      // We will want the flux in W/m2 for the coupling so no face integral for now,
+      // this can cause issues if we start using face area in the kernels
+      // for more than just face integral multipliers.
+      // Also, if we decide to no require overlapping meshes on the boundary
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]:
```suggestion
      // Also, if we decide to not require overlapping meshes on the boundary
```

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 29]
```diff
@@ -0,0 +1,417 @@
...
+      {},
+      "The interfaces where we would like to add conjugate heat transfer handling.");
+
+  params.addRangeCheckedParam<unsigned int>(
+      "max_cht_fpi",
+      1,
+      "max_cht_fpi >= 1",
+      "Number of maximum fixed point iterations (FPI). Currently only applied to"
+      " conjugate heat transfer simulations. The default value of 1 essentially keeps"
+      " the FPI feature turned off.");
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]:
```suggestion
      " the FPI feature turned off. CHT iteration ends after this number of iteration even if the tolerance is not met.");
```

File: modules/navier_stokes/src/utils/CHTHandler.C [hunk 30]
```diff
@@ -0,0 +1,417 @@
...
+
+  params.addRangeCheckedParam<Real>(
+      "cht_heat_flux_tolerance",
+      1e-5,
+      "cht_heat_flux_tolerance > 0",
+      "The relative tolerance for terminating conjugate heat transfer iteration.");
```
Comment (GiudGiud) [inline modules/navier_stokes/src/utils/CHTHandler.C]:
```suggestion
      "The relative tolerance for terminating conjugate heat transfer iteration before the maximum number of CHT iterations. Relative tolerance is ignore if the maximum number of CHT iterations is reached.");
```

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/gold/cht_neu-dir_out_y_vs_tf_0001.csv (hunk 14 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/gold/cht_neu-dir_out_y_vs_tf_0001.csv:3]: It is the relative line width, makes it nice because we can compares the same problem with different orientation.

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests (hunk 15 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests]: Great idea! I'll try it out!

File: modules/navier_stokes/src/linearfvbcs/LinearFVRobinCHTBC.C (hunk 22 above)
Comment (grmnptr) [reply modules/navier_stokes/src/linearfvbcs/LinearFVRobinCHTBC.C]: Oh yes, great catch!

File: modules/navier_stokes/include/base/NS.h (hunk 7 above)
Comment (grmnptr) [reply modules/navier_stokes/include/base/NS.h:192]: I was referring to the difference between 
```
enum something;
```
and 
```
enum class something;
```
The former can be interpreted as an int.

File: modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests (hunk 15 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/cht/conjugate_heat_transfer/tests]: Worked out

**Review by GiudGiud — APPROVED** [review]
please consider contributing a verification / validation benchmark to OpenPGH
